Configurable glob

bstaudacher opened the pull request:
### The problem

Per [the readme file](https://github.com/Shopify/erb-lint/blob/3a2ee22877d2361ff2591b4f17a52430eecd496e/README.md#usage):

> `erblint --lint-all --enable-all-linters` will run all available linters on all ERB files in the current directory or its descendants `(**/*.html{+*,}.erb)`.

Unfortunately this glob does not encompass all ERB files for some projects. Notably, `.text.erb` and `.js.erb` files will not be linted.

### The culprit
[The relevant code](https://github.com/Shopify/erb-lint/blob/3a2ee22877d2361ff2591b4f17a52430eecd496e/lib/erb_lint/cli.rb#L14) is found in `cli.rb`.

The glob is defined as a class constant with no configuration possible.

### The solution
This PR changes the class constant to a private method. The private method first attempts to load a glob from [the configuration file](https://github.com/Shopify/erb-lint/tree/3a2ee22877d2361ff2591b4f17a52430eecd496e#configuration), if that fails it falls back on the existing default of `**/*.html{+*,}.erb`.

kevinhughes27 commented:
ping! @EiNSTeiN- would it be possible for someone to have a look at this PR?

bstaudacher commented:
@cjhutchi any feedback on this guy? Would love to know if this could be added in or not.

thegedge commented on lib/erb_lint/cli.rb:
```diff
@@ -168,18 +168,27 @@ def load_options(args)
+        config = RunnerConfig.new(file_loader.yaml(config_filename), file_loader)
+        @config = RunnerConfig.default.merge(config)
+        return @config.to_hash["glob"]
+      end
```
At the point that `glob` is called, `load_config` has already been called (in `initialize`), so no need to load the config again.

bstaudacher replied:
I had originally semi-redone the config load as `glob` is used [during the options parse](https://github.com/Shopify/erb-lint/pull/143/files#diff-a989663d2711ff35a3a5fc5aa185f838R262) which happens before `load_config`.

Another option I explored was moving the load_config call to run before opts, however this has knock on effects.

Namely, the LinterRegistry loads custom linters into @options and this is necessary for `runner_config_override` to function properly.

I have a fix I am working on that will make this more readable, please wait before merge.

bstaudacher commented:
I am seeing strange failures on CI that I am not able to reproduce locally.
I made a test branch that PR'd my previously passing commits and it now fails with the same message.
Any ideas on how to resolve this? The error for reference:

```
     RuboCop::ValidationError:
       The `Layout/AlignArguments` cop has been renamed to `Layout/ArgumentAlignment`.
       (obsolete configuration found in .erblint-rubocop20191204-16552-1p69s5q, please update it)
```

The test PR branch:

https://github.com/Shopify/erb-lint/pull/144

cjhutchi commented:
> I am seeing strange failures on CI that I am not able to reproduce locally.
> I made a test branch that PR'd my previously passing commits and it now fails with the same message.
> Any ideas on how to resolve this? The error for reference:
> 
> ```
>      RuboCop::ValidationError:
>        The `Layout/AlignArguments` cop has been renamed to `Layout/ArgumentAlignment`.
>        (obsolete configuration found in .erblint-rubocop20191204-16552-1p69s5q, please update it)
> ```
> 
> The test PR branch:
> 
> #144

@bstaudacher never seen that before ¯\_(ツ)_/¯. Sounds like there's a reference to an outdated cop somewhere in the config perhaps?

exterm commented:
@bstaudacher the master branch was actually broken. We've since fixed that. A rebase should fix CI on this branch

bstaudacher commented:
@exterm, rebased and added in a better test for good measure.

Thank you for getting around to fixing the master branch!

Let me know if there is anything else I should address on this PR.

exterm commented:
@thegedge should take another look once he is back, but it looks ready to go from my perspective. Thank you for the contribution.

thegedge commented on lib/erb_lint/cli.rb:
```diff
@@ -33,11 +33,12 @@ def initialize
       @files = dupped_args
 
-      load_config
+      merge_config
```
How do you feel about merging this with `load_config`? It feels like it should be "atomic".

My main concern is that the separation makes it feel like they're two separate things. On top of that, putting `load_config` at the beginning of this function makes it seem like it's important for `load_config` to be called before `load_options`.

Another option is to keep the `merge_config` method, but put it at the bottom of `load_config`. That ensures correct ordering 🙂

bstaudacher replied:
It is interesting - this is a tough one to defend, but after some pretty serious debugging I feel that the loading, gathering of options, and then merging of the config are all quite separate.

On master currently, the options are gathered, then config is loaded and merged:

Options loading + config loading:
https://github.com/Shopify/erb-lint/blob/44b4b9629872bc952f76ea9f5a99fa738d0b832c/lib/erb_lint/cli.rb#L35-L40

Config merging within the load method:
https://github.com/Shopify/erb-lint/blob/44b4b9629872bc952f76ea9f5a99fa738d0b832c/lib/erb_lint/cli.rb#L150-L161

The problem I have is that now that the glob can be configured, this line needs to be changed:
https://github.com/Shopify/erb-lint/blob/44b4b9629872bc952f76ea9f5a99fa738d0b832c/lib/erb_lint/cli.rb#L256-L258

To accurately report a custom glob, I need to have the configuration loaded.
Originally I thought, oh no problem - move the `load_config` call to before the options. The problem is that the merge step inside of `load_config` can rely on options that have been passed (linters enabled etc.).

Due to the above ordering problem, I split the functions out and rearranged them.

I am absolutely open to ways to get around this, perhaps there is a better way to report the custom glob in the `option_parser`?

Thank you for your time, I realize the above is quite in depth.

thegedge replied:
Really appreciate the thorough response, @bstaudacher. As I guessed, there _is_ an underlying reason to the ordering, it's just a little difficult to make the connection.

Now that I know about the issue, I may follow up to address this, but I really want to push your contribution through because it's super useful. I think the way to go about it is to just not show the glob in the option's, but rephrase the help in some way to mention the glob.

Again, thanks for the thoughtful response, and we really appreciate the contribution 🙌

mreith101 replied:
The error that you were getting earlier related to the RuboCop issue incase you never figured it out, likely suggests that the YAML config file for RuboCop has had syntax changed, so in this case, the rule you have set for: **Layout/AlignArguments** has had its name changed from that to this: **Layout/ArgumentAlignment.**  You can see that they are similar but the name is slightly different so the proposed solution could have been to update the rubocop configuration to refer to the correct rule you want to enable.

 RuboCop::ValidationError:
       The `Layout/AlignArguments` cop has been renamed to `Layout/ArgumentAlignment`.
       (obsolete configuration found in .erblint-rubocop20191204-16552-1p69s5q, please update it)

Figured I would chime in... :)

thegedge commented:
I'll take care of merging and releasing a new version tomorrow. Thanks so much, @bstaudacher! ⭐️
